bundles/org.openhab.binding.irobot/src/main/java/org/openhab/binding/irobot/internal/handler/RoombaHandler.java

```diff
@@ -205,178 +269,95 @@ private void sendSchedule(MQTTProtocol.Schedule schedule) {
-            this.connection = null;
+        credentialRequester = null;
+        if (UNKNOWN.equals(config.getBlid()) || UNKNOWN.equals(config.getPassword())) {
+            credentialRequester = scheduler.schedule(this::getCredentials, 10000, TimeUnit.MILLISECONDS);
+        } else {
+            scheduler.execute(this::connect);
         }
     }
 
-    private void scheduleReconnect() {
-        reconnectReq = scheduler.schedule(this::connect, RECONNECT_DELAY_SEC, TimeUnit.SECONDS);
-    }
-
-    public void onConnected() {
-        updateStatus(ThingStatus.ONLINE);
+    // In order not to mess up our connection state we need to make sure that connect()
+    // and disconnect() are never running concurrently, so they are synchronized
+    private synchronized void connect() {
+        IRobotConfiguration config = getConfigAs(IRobotConfiguration.class);
+        final String address = config.getAddress();
+        logger.debug("Connecting to {}", address);
+
+        final String blid = config.getBlid();
+        final String password = config.getPassword();
+        if (UNKNOWN.equals(blid) || UNKNOWN.equals(password)) {
+            final String message = "Robot authentication is required";
+            updateStatus(OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, message);
+            scheduler.execute(this::getCredentials);
+        } else {
+            final String message = "Robot authentication is successful";
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.CONFIGURATION_PENDING, message);
+            connection.connect(address, blid, password);
+        }
     }
 
-    @Override
-    public void processMessage(String topic, byte[] payload) {
-        String jsonStr = new String(payload);
+    public void receive(final String topic, final String json) {
         MQTTProtocol.StateMessage msg;
 
-        logger.trace("Got topic {} data {}", topic, jsonStr);
+        logger.trace("Got topic {} data {}", topic, json);
 
         try {
             // We are not consuming all the fields, so we have to create the reader explicitly
             // If we use fromJson(String) or fromJson(java.util.reader), it will throw
             // "JSON not fully consumed" exception, because not all the reader's content has been
             // used up. We want to avoid that also for compatibility reasons because newer iRobot
             // versions may add fields.
-            JsonReader jsonReader = new JsonReader(new StringReader(jsonStr));
+            JsonReader jsonReader = new JsonReader(new StringReader(json));
             msg = gson.fromJson(jsonReader, MQTTProtocol.StateMessage.class);
-        } catch (JsonParseException e) {
-            logger.warn("Failed to parse JSON message from {}: {}", config.ipaddress, e.toString());
-            logger.warn("Raw contents: {}", payload);
+        } catch (JsonParseException exception) {
+            IRobotConfiguration config = getConfigAs(IRobotConfiguration.class);
```

Comment:
Why is the configuration read again here? Actually it should be sufficient to read it once in `initialize()`.

Reply:
removed usage of config. Using Thing UID instead